zebra-consensus/src/transaction/tests.rs

```diff
@@ -62,19 +37,82 @@ fn v5_fake_transactions_for_network(network: Network) -> Result<(), Report> {
                     ..
                 } => {
                     if let Some(s) = sapling_shielded_data {
-                        super::check::sapling_balances_match(&s)?;
+                        check::sapling_balances_match(&s)?;
 
                         for spend in s.spends_per_anchor() {
-                            super::check::spend_cv_rk_not_small_order(&spend)?
+                            check::spend_cv_rk_not_small_order(&spend)?
                         }
                         for output in s.outputs() {
-                            super::check::output_cv_epk_not_small_order(&output)?;
+                            check::output_cv_epk_not_small_order(&output)?;
                         }
                     }
                 }
                 _ => panic!("we should have no tx other than 5"),
             }
         }
     }
+
     Ok(())
 }
+
+#[test]
+fn v5_transaction_with_no_inputs_fails_validation() {
+    let transaction = v5_fake_transactions_for_network(
+        Network::Mainnet,
+        zebra_test::vectors::MAINNET_BLOCKS.iter(),
+    )
+    .rev()
+    .find(|transaction| {
+        transaction.inputs().is_empty()
+            && transaction.sapling_spends_per_anchor().next().is_none()
+            && transaction.orchard_actions().next().is_none()
+            && transaction.joinsplit_count() == 0
+            && (!transaction.outputs().is_empty() || transaction.sapling_outputs().next().is_some())
+    })
+    .expect("At least one transaction with no inputs in the test vectors");
+
+    assert_eq!(
+        check::has_inputs_and_outputs(&transaction),
+        Err(TransactionError::NoInputs)
+    );
+}
+
+#[test]
+fn v5_transaction_with_no_outputs_fails_validation() {
+    let transaction = v5_fake_transactions_for_network(
+        Network::Mainnet,
+        zebra_test::vectors::MAINNET_BLOCKS.iter(),
+    )
+    .rev()
+    .find(|transaction| {
+        transaction.outputs().is_empty()
+            && transaction.sapling_outputs().next().is_none()
+            && transaction.orchard_actions().next().is_none()
+            && transaction.joinsplit_count() == 0
+            && (!transaction.inputs().is_empty()
+                || transaction.sapling_spends_per_anchor().next().is_some())
+    })
+    .expect("At least one transaction with no outputs in the test vectors");
+
+    assert_eq!(
+        check::has_inputs_and_outputs(&transaction),
+        Err(TransactionError::NoOutputs)
+    );
+}
+
+fn v5_fake_transactions_for_network<'b>(
+    network: Network,
+    blocks: impl DoubleEndedIterator<Item = (&'b u32, &'b &'static [u8])> + 'b,
```

Comment:
In production code I would have used `.cloned()` to simplify the lifetimes in this signature, but we're pretty relaxed about test code.

Reply:
Hmm, I just tried `.cloned()` now, but it didn't work immediately. The problem is that the type is `&(&u32, &&[u8])` so it became `(&u32, &&[u8])`. I think I'd have to manually destructure the tuple and then deref each element :thinking: But I'll leave it as is, to not make it more complicated :sweat_smile:

Reply:
You might want `Borrow::to_owned` here - or you might want to clone much earlier in the process.
https://doc.rust-lang.org/std/borrow/trait.ToOwned.html#tymethod.to_owned

But this is ok for now.